src/main/java/com/google/javascript/clutz/DeclarationGenerator.java

```diff
@@ -486,18 +483,24 @@ private void processUnprovidedTypes(Set<String> provides) {
       // AFAICT, there is no api for going from type to symbol, so iterate all symbols first.
       for (TypedVar symbol : compiler.getTopScope().getAllSymbols()) {
         String name = symbol.getName();
-        // skip unused symbols, symbols already emitted or symbols whose namespace is emitted.
+        String namespace = getNamespace(name);
+        // skip unused symbols, symbols already emitted or symbols whose namespace is emitted
+        // (unless the symbols have their own provide).
         if (!typesUsed.contains(name) || typesEmitted.contains(name) ||
-            typesEmitted.contains(getNamespace(name))) {
+            (!transitiveProvides.contains(name) && typesEmitted.contains(namespace))) {
```

Comment:
I have hunch that this will need even more iteration in the future, but I will leave it for a refactoring after we know clutz with strict deps works. The core issue is that at two different places we do the '(TypedVar symbol : compiler.getTopScope().getAllSymbols())' and decide whether to emit symbol or not, and the logic is not shared.